### ISSUE TITLE
Allowing only plugins to run, and allowing options for them

### ISSUE THREAD

**Issue by anthonyalayo** [ascopes/protobuf-maven-plugin#168]
At the moment, there are two missing features from this project:

1. Allowing an execution stage to run _only_ with a plugin, and no language base.
2. Support for plugin options

I made a working prototype of the idea here:
https://github.com/ascopes/protobuf-maven-plugin/pull/167

Could we discuss iterations to make this clear for check-in?

Thanks,
Anthony

**Comment by ascopes** [comment]
Hey, thanks for raising this.

For the first point, all languages can be toggled off if Java is not desired. You should just be able to set `<javaEnabled>false</javaEnabled>`, does that work for your use case or is there other issues that you've had trying this?

For the second point, I totally agree, this is needed. Good spot! Happy to include it.

**Comment by ascopes** [comment]
Responded on the PR you raised...

> I've been thinking about this and I think that in the long run, it is going to be more beneficial to make a breaking change for this so that the plugin attributes in the config can take an options parameter each. This allows directing options to individual plugins rather than all plugins.

> This would be breaking purely on the basis that I'd have to slightly change how URL and Path based plugins get provided in config.

> I have some code locally that I can push later once I have tidied it up a bit, I will run that past you once it is ready.

My plan is to:

- add an optional `options` attribute to the `binaryMavenPlugins` parameter
- add an optional `options` attribute and `jvmMavenPlugins` parameter
- change the `binaryPathPlugins` `List<String>` parameter to a list of objects that have a `name` and optional `options` attribute.
- change the `binaryUrlPlugins` `List<URL>` parameter to a list of objects that have a `url` and optional `options` parameter.

Using objects for the path and url plugin params will be a bit more verbose but it has the added benefit of allowing the addition of further attributes in the future without being a breaking change.

**Comment by ascopes** [comment]
Would it be possible to provide an example of a plugin that consumes an option and an expectation of what you expect to happen? I can then link it to this issue as an integration test to ensure I cover your specific use case.

**Comment by anthonyalayo** [comment]
@ascopes here's an example:
https://github.com/alpancs/protoc-gen-pubsub-schema

**Comment by ascopes** [comment]
Thanks, I've released this under 2.0.0-alpha1. Please can you test that this covers this use case correctly for you on your side when you get the time?

**Comment by anthonyalayo** [comment]
@ascopes it worked like a charm! Here's my diff for context:

```
                   <configuration>
-                    <pluginOnly>true</pluginOnly>
-                    <pluginOption>schema-syntax=proto3</pluginOption>
+                    <javaEnabled>false</javaEnabled>
                     <registerAsCompilationRoot>false</registerAsCompilationRoot>
                     <binaryUrlPlugins>
                       <binaryUrlPlugin>
-                        file://${env.HOME}/go/bin/protoc-gen-pubsub-schema
+                        <url>file://${env.HOME}/go/bin/protoc-gen-pubsub-schema</url>
+                        <options>schema-syntax=proto3</options>
                       </binaryUrlPlugin>
                     </binaryUrlPlugins>
```

**Comment by ascopes** [comment]
might be the text outside the tags in the binsryUrlPlugin block

**Comment by anthonyalayo** [comment]
@ascopes that line is removed (check the plus and minus on the left)

**Comment by ascopes** [comment]
ahh sorry! That will teach me to respond to GitHub comments immediately after waking up!

I'm not entirely sure in that case. It may be an IntelliJ bug. I've spotted similar issues in the past with other plugins.

Might be worth raising a YouTrack bug with them?

**Comment by anthonyalayo** [comment]
No worries! Perhaps some sort of feature of plugins, If I spot it I'll let you know.

**Comment by ascopes** [comment]
Closing as the main functionality is present for this in 2.0.0-alpha1 now and the rest of the request is now in GH-182.

**Comment by ascopes** [comment]
Released under v2.0.0 as stable.
